_posts/2021-02-10-native-k8s-with-ha.md

```diff
@@ -0,0 +1,112 @@
+ 2. The active leader publishes its address to the ConfigMap. It’s important to  note that Flink uses the same ConfigMap for contending lock and storing the leader address. This ensures that there is no unexpected change snuck in during a periodic update.
+ 3. The leader retrieval service is used to find the active leader’s address and allow the components to then register themselves. For example, TaskManagers retrieve the address of ResourceManager and JobManager for the registration and offering slots. Flink uses a Kubernetes watch in the leader retrieval service — once the content of ConfigMap changes, it usually means that the leader has changed, and so the listener can get the latest leader address immediately.
+ 4. All other meta information (e.g. running jobs, job graphs, completed checkpoints and checkpointer counter) will be directly stored in the corresponding ConfigMaps. Only the leader can update the ConfigMap. The HA data will only be cleaned up once the Flink cluster reaches the global terminal state. Please note that only the pointers are stored in the ConfigMap; the concrete data has been stored in DistributedStorage. This level of indirection is necessary to keep the amount of data in ConfigMap small (ConfigMap is built for data less than 1MB whereas state can grow to multiple GBs).
+
+# Example: Application cluster with HA
+
+Please note that you need a running Kubernetes cluster and to get kube config properly set to follow along.
+You can use `kubectl get nodes` for verifying that you’re all set.
+In this blog post, we’re using [minikube](https://minikube.sigs.k8s.io/docs/start/) for local testing.
+
+1. Build a Docker image with the Flink job (my-flink-job.jar) baked in
+
+        FROM flink:1.12.1
+        RUN mkdir -p $FLINK_HOME/usrlib
+        COPY /path/of/my-flink-job.jar $FLINK_HOME/usrlib/my-flink-job.jar
+
+    Use the above Dockerfile to build a user image and then push it to your remote image repository.
+
+2. Start a Flink application cluster
+
+        $ ./bin/flink run-application -d -p 4 -t kubernetes-application \
+            -Dkubernetes.cluster-id=k8s-ha-app-1 \
+            -Dkubernetes.container.image=<user-image> \
+            -Dkubernetes.jobmanager.cpu=0.5 -Dkubernetes.taskmanager.cpu=0.5 \
+            -Dtaskmanager.numberOfTaskSlots=4 \
+            -Dkubernetes.rest-service.exposed.type=NodePort \
+            -Dhigh-availability=org.apache.flink.kubernetes.highavailability.KubernetesHaServicesFactory \
+            -Dhigh-availability.storageDir=s3://flink-bucket/flink-ha \
+            -Drestart-strategy=fixed-delay -Drestart-strategy.fixed-delay.attempts=10 \
+            -Dcontainerized.master.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-1.12.1.jar \
+            -Dcontainerized.taskmanager.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-1.12.1.jar \
+            local:///opt/flink/usrlib/my-flink-job.jar
```

Comment:
```suggestion
        $ ./bin/flink run-application \
            --detached \
            --parallelism 4 \
            --target kubernetes-application \
            -Dkubernetes.cluster-id=k8s-ha-app-1 \
            -Dkubernetes.container.image=<user-image> \
            -Dkubernetes.jobmanager.cpu=0.5 \
            -Dkubernetes.taskmanager.cpu=0.5 \
            -Dtaskmanager.numberOfTaskSlots=4 \
            -Dkubernetes.rest-service.exposed.type=NodePort \
            -Dhigh-availability=org.apache.flink.kubernetes.highavailability.KubernetesHaServicesFactory \
            -Dhigh-availability.storageDir=s3://flink-bucket/flink-ha \
            -Drestart-strategy=fixed-delay \
            -Drestart-strategy.fixed-delay.attempts=10 \
            -Dcontainerized.master.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-1.12.1.jar \
            -Dcontainerized.taskmanager.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-1.12.1.jar \
            local:///opt/flink/usrlib/my-flink-job.jar
```

In the deployment docs we agreed on using long parameters. Another thing we tried to follow is to put one parameter per line. We hope that this improves the readability of the command. We could do that in the blog post as well, if you agree.